README.md
# e_lib
e_xtended basic library for common requirements of my c projects

## Project License
Copyright (c) 2023, [Diogo Ferreira](https://github.com/diogoefl)

SPDX-License-Identifier: [BSD-3-Clause](https://spdx.org/licenses/BSD-3-Clause.html)

See [license](./LICENSE) at this project root for more detailed information

## Build system used for testing

- CMake 3.23+
- LLVM 16+
- Ninja 1.11+
- C23

## Planned features

- e_string_t: string non-based on NUL terminator
- e_bigdec_t: simple arbitrary precision decimal
- e_bigint_t: simple arbitrary precision integer
- e_vector_t: vector implementation 
- e_format: simple string formatting system

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## Project License
Copyright (c) 2023, [Diogo Ferreira](https://github.com/diogoefl)

SPDX-License-Identifier: [BSD-3-Clause](https://spdx.org/licenses/BSD-3-Clause.html)

See [license](./LICENSE) at this project root for more detailed information


## [Unreleased]

### To do

- Remove main executable and keep only testing for library
- Add example for usage
- Implement rest of e_string_t
- Implement e_bigdec_t
- Implement e_bigint_t
- Implement e_vector_t
- Implement e_format
- Reset std to C17

### Added

- e_string_t struct
- e_string_validate buffer function
- e_string_valdiate buffer unit testing
- Created base repository with simple guidelines
- Adjusting template project for library usage

[0.0.1]: https://github.com/diogoefl/cmake_c_project_template/releases/tag/v0.0.1

test/e_lib_add9_test.c
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* Project example lib add9 test */

#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

#include "e_lib.h"

void test_1(void)
{
    fprintf(stdout, "%s ... ", u8"[e_lib_test] Testing e_add9 with 10");
    
    if (e_add9(10) == 19) {
        fprintf(stdout, "%s\n", u8"SUCCESS");
    } else {
        fprintf(stdout, "%s\n", u8"FAIL");
        exit(EXIT_FAILURE);
    }
}

void test_2(void)
{
    fprintf(stdout, "%s ... ", u8"[e_lib_test] Testing e_add9 with 868");
    
    if (e_add9(868) == 877) {
        fprintf(stdout, "%s\n", u8"SUCCESS");
    } else {
        fprintf(stdout, "%s\n", u8"FAIL");
        exit(EXIT_FAILURE);
    }
}

int main(void)
{
    test_1();
    test_2();
    return EXIT_SUCCESS;
}

test/e_string/e_string_validate_test.c
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* e_string_validate namespace testing*/

#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <locale.h>
#include <string.h>

#include "e_string.h"

/* private function priv_process_args
 *
 * retrieve arguments given and process data to create e_string_t
 */
e_string_t priv_process_args(const int argc, char*const argv[])
{
    if (argc < 3) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: missing argument for testing",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    } else if (argc > 3) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: too many arguments for testing",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    }

    const char* input = argv[1];
    int expected_size = atoi(argv[2]);

    if (expected_size <= 0) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: expected positive number greater than 0 as second argument",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    }
    fprintf(stdout, "%s: %d\n",
            u8"[e_string_validate] input for expected size", expected_size);

    const size_t input_length = strlen(input);
    const size_t memory_buffer = expected_size * sizeof(uint8_t);

    fprintf(stdout, "%s: %s\n",
            u8"[e_string_validate] input given", input);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] input length", input_length);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] input memory required", memory_buffer);
    
    e_string_t result = {
        .data_length = input_length,
        .buffer_capacity = memory_buffer,
        .data = malloc(memory_buffer)
    };

    for (size_t i = 0; i < expected_size; i++) {
        result.data[i] = input[i];
    }
    
    return result;
}

/* private function priv_validate_buffer
 *
 * check if buffer_capacity and data_length are compatible
 */
bool priv_validate_buffer(e_string_t string)
{
    if (string.buffer_capacity < string.data_length) {
        fprintf(stdout, "%s %d\n%s\n",
                u8"[e_string_validate] error: corrupted e_string may return e_string_errno_t of",
                E_STRING_INVALID_BUFFER,
                u8"[e_string_validate] error: memory buffer is incompatible with data length");
    } else if (string.buffer_capacity == string.data_length) {
        fprintf(stdout, "%s\n",
                u8"[e_string_validate] valid: memory buffer is equal data length");
    } else {
        fprintf(stdout, "%s\n",
                u8"[e_string_validate] valid: memory buffer is bigger than data length");
    }
     fprintf(stdout, "%s %zu\n%s %zu\n",
             u8"[e_string_validate] buffer detected:", string.buffer_capacity,
             u8"[e_string_validate] length required:", string.data_length);

     return (e_string_validate(&string) == E_STRING_INVALID_BUFFER) ? false : true;
}


/* e_string_validate_test
 *
 * create string based on input and test if string are valid on its memory,
 * and also if all content is UTF-8 compliant
 */
int main(int argc, char* argv[])
{
    /* process input */
    setlocale(LC_ALL, "en_US.utf8");
    e_string_t string = priv_process_args(argc, argv);
    
    /* execute validations */
    if (priv_validate_buffer(string) == false) {
        exit(EXIT_FAILURE);
    } else if (e_string_validate(&string) != E_STRING_SUCCESS) {
        exit(EXIT_FAILURE);
    }

    /* no errors found - print string data */
    fprintf(stdout, "%s: %.*s\n",
            u8"[e_string_validate] e_string data", (int)string.data_length, string.data);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] e_string length", string.data_length);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] e_string buffer size", string.buffer_capacity);
    

    return EXIT_SUCCESS;
}

test/e_string/e_string_validate_buffer_test.c
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* e_string_validate namespace testing*/

#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <locale.h>
#include <string.h>

#include "e_string.h"

/* private function priv_process_args
 *
 * retrieve arguments given and process data to create e_string_t
 */
e_string_t priv_process_args(const int argc, char*const argv[])
{
    if (argc < 3) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: missing argument for testing",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    } else if (argc > 3) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: too many arguments for testing",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    }

    const char* input = argv[1];
    int expected_size = atoi(argv[2]);

    if (expected_size <= 0) {
        fprintf(stdout, "%s\n%s\n",
                u8"[e_string_validate] error: expected positive number greater than 0 as second argument",
                u8"[e_string_validate] usage: e_string_validate_test string length_number");
        exit(EXIT_FAILURE);
    }
    fprintf(stdout, "%s: %d\n",
            u8"[e_string_validate] input for expected size", expected_size);

    const size_t input_length = strlen(input);
    const size_t memory_buffer = expected_size * sizeof(uint8_t);

    fprintf(stdout, "%s: %s\n",
            u8"[e_string_validate] input given", input);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] input length", input_length);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] input memory required", memory_buffer);
    
    e_string_t result = {
        .data_length = input_length,
        .buffer_capacity = memory_buffer,
        .data = malloc(memory_buffer)
    };

    for (size_t i = 0; i < expected_size; i++) {
        result.data[i] = input[i];
    }
    
    return result;
}

/* private function priv_validate_buffer
 *
 * check if buffer_capacity and data_length are compatible
 */
bool priv_validate_buffer(e_string_t string)
{
    if (string.buffer_capacity < string.data_length) {
        fprintf(stdout, "%s %d\n%s\n",
                u8"[e_string_validate] error: corrupted e_string may return e_string_errno_t of",
                E_STRING_INVALID_BUFFER,
                u8"[e_string_validate] error: memory buffer is incompatible with data length");
    } else if (string.buffer_capacity == string.data_length) {
        fprintf(stdout, "%s\n",
                u8"[e_string_validate] valid: memory buffer is equal data length");
    } else {
        fprintf(stdout, "%s\n",
                u8"[e_string_validate] valid: memory buffer is bigger than data length");
    }
     fprintf(stdout, "%s %zu\n%s %zu\n",
             u8"[e_string_validate] buffer detected:", string.buffer_capacity,
             u8"[e_string_validate] length required:", string.data_length);

     return (e_string_validate(&string) == E_STRING_INVALID_BUFFER) ? false : true;
}


/* e_string_validate_buffer_test_valid
 *
 * create string based on input and test if string are valid.
 */
int main(int argc, char* argv[])
{
    /* process input */
    setlocale(LC_ALL, "en_US.utf8");
    e_string_t string = priv_process_args(argc, argv);
    
    /* execute validations */
    if (priv_validate_buffer(string) == false) {
        exit(EXIT_FAILURE);
    }

    /* no errors found - print string data */
    fprintf(stdout, "%s: %.*s\n",
            u8"[e_string_validate] e_string data", (int)string.data_length, string.data);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] e_string length", string.data_length);
    fprintf(stdout, "%s: %zu\n",
            u8"[e_string_validate] e_string buffer size", string.buffer_capacity);
    

    return EXIT_SUCCESS;
}

include/e_string_utf8.h
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* e_string_utf8 helper header 
 *
 * this is a helper module not exported.
 * final user is not supposed to use this header.
 */

#ifndef E_STRING_UTF8_H
#define E_STRING_UTF8_H

#include <stdbool.h>
#include <stdint.h>

/* e_string_utf8_is_ascii
 *
 * checks if data is compliant to US ASCII on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * US ASCII is represented in e_string_t as 1 byte (8 bit) with the value
 * ranging from 0x00 upto 0x7F. 
 *
 * if sanity check is true, the following items from 0x00 to 0x1F are considered
 * invalid, as they can represent conflicts for usage and should not be
 * inside of a e_string_t which attempts to represent simple text strings.
 * if you need a string with this type of characters, please use char* instead.
 *   - 0x00 NUL
 *   - 0x01 SOH
 *   - 0x02 STX
 *   - 0x03 ETX
 *   - 0x04 EOT
 *   - 0x05 ENT
 *   - 0x06 ACK
 *   - 0x07 \a
 *   - 0x08 \b
 *   - 0x0B VT
 *   - 0x0C \f
 *   - 0x0E SO
 *   - 0x0F SI
 *   - 0x10 DLE
 *   - 0x11 DC1
 *   - 0x12 DC2
 *   - 0x13 DC3
 *   - 0x14 DC4
 *   - 0x15 NAK
 *   - 0x16 SYN
 *   - 0x17 ETB
 *   - 0x18 CAN
 *   - 0x19 EM
 *   - 0x1A SUB
 *   - 0x1B \e
 *   - 0x1C FS
 *   - 0x1D GS
 *   - 0x1E RS
 *   - 0x1F US
 *   - 0x7F DEL
 *
 * ranging from 0x00 to 0x1F, the following items are valid for e_string_t,
 * regardless of sanity check value.
 *   - 0x09 \t
 *   - 0x0A LF
 *   - 0x0D CR
 */
bool e_string_utf8_is_ascii(const uint8_t u8_byte, const bool sanity);

/* e_string_utf8_is_non_overlong
 *
 * checks if data is compliant to non-overlong 2 bytes on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * non-overlong is represented in e_string_t as 2 bytes (16 bit) with the value
 * of the first byte ranging from 0xC2 upto 0xDF and the second byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is non-overlong you need both bytes for the correct output.
 */
bool e_string_utf8_is_non_overlong(const uint8_t u8_1st_byte,
                                   const uint8_t u8_2nd_byte);


/* e_string_utf8_is_excluding_overlong
 *
 * checks if data is compliant to excluding overlong 3 bytes on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * excluding overlong is represented in e_string_t as 3 bytes (24 bit) with the
 * value of the first byte being 0xE0, the second byte ranging from 0xA0 upto
 * 0xBF and the third byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is excluding overlong you need all bytes for the correct output.
 */
bool e_string_utf8_is_excluding_overlong(const uint8_t u8_1st_byte,
                                         const uint8_t u8_2nd_byte,
                                         const uint8_t u8_3rd_byte);


/* e_string_utf8_is_straight_3byte
 *
 * checks if data is compliant to straight 3-byte on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * straight 3-byte is represented in e_string_t as 3 bytes (24 bit) with the
 * value of the first byte ranging from 0xE1 upto 0xEC or from 0xEE upto 0xEF,
 * the second byte ranging from 0x80 upto 0xBF and the third byte ranging from
 * 0x80 upto 0xBF.
 *
 * if the codepoint is straight 3-byte you need all bytes for the correct output.
 */
bool e_string_utf8_is_straight_3byte(const uint8_t u8_1st_byte,
                                     const uint8_t u8_2nd_byte,
                                     const uint8_t u8_3rd_byte);


/* e_string_utf8_is_excluding_surrogates
 *
 * checks if data is compliant to excluding surrogates on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * excluding surrogates is represented in e_string_t as 3 bytes (24 bit) with 
 * the value of the first byte is 0xED, the second byte ranging from 0x80 upto
 * 0x9F and the third byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is excluding surrogates you need all bytes for the correct output.
 */
bool e_string_utf8_is_excluding_surrogates(const uint8_t u8_1st_byte,
                                           const uint8_t u8_2nd_byte,
                                           const uint8_t u8_3rd_byte);


/* e_string_utf8_is_plane_1to3
 *
 * checks if data is compliant to planes 1-3 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * planes 1-3 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte is 0xF0, the second byte ranging from 0x90 upto
 * 0xBF, the third byte ranging from 0x80 upto 0xBF and the fourth byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is planes 1-3 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_1to3(const uint8_t u8_1st_byte,
                                 const uint8_t u8_2nd_byte,
                                 const uint8_t u8_3rd_byte,
                                 const uint8_t u8_4th_byte);


/* e_string_utf8_is_plane_4to15
 *
 * checks if data is compliant to planes 4-15 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * planes 4-15 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte ranging from 0xF1 upto 0xF3, the second byte
 * ranging from 0x80 upto 0xBF, the third byte ranging from 0x80 upto 0xBF and
 * the fourth byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is planes 4-15 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_4to15(const uint8_t u8_1st_byte,
                                  const uint8_t u8_2nd_byte,
                                  const uint8_t u8_3rd_byte,
                                  const uint8_t u8_4th_byte);


/* e_string_utf8_is_plane_16
 *
 * checks if data is compliant to plane 16 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * plane 16 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte is 0xF4, the second byte ranging from 0x80 upto
 * 0x8F, the third byte ranging from 0x80 upto 0xBF and the fourth byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is plane 16 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_16(const uint8_t u8_1st_byte,
                               const uint8_t u8_2nd_byte,
                               const uint8_t u8_3rd_byte,
                               const uint8_t u8_4th_byte);

#endif /* E_STRING_UTF8_H */

src/e_string/e_string_from.c
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* e_string_from implementation
 *
 * this module implements constructors of e_string_t.
 * 
 * usage: add #include "e_string.h" to your file and link to e_string library
 */

#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "e_string.h"


e_string_t e_string_from_cstr(const char* cstr)
{

    const size_t input_length = strlen(cstr);
    const size_t memory_buffer = input_length * sizeof(uint8_t);
    
    e_string_t result = {
        .data_length = input_length,
        .buffer_capacity = memory_buffer,
        .data = malloc(memory_buffer)
    };

    for (size_t i = 0; i < input_length; i++) {
        result.data[i] = cstr[i];
    }
    
    return result;
}


e_string_t e_string_from_uint64(const uint64_t number)
{
    /* define how many digits there are in number */
    const int digits = (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%ju", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_uint32(const uint32_t number)
{
    /* define how many digits there are in number */
    const int digits = (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%u", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_uint16(const uint16_t number)
{
    /* define how many digits there are in number */
    const int digits = (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%hu", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_uint8(const uint8_t number)
{
    /* define how many digits there are in number */
    const int digits = (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%hu", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_int64(const int64_t number)
{
    /* define how many digits there are in number */
    const int digits = (number < 0) ? (int)(log10(number) + 2)
                                    : (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%jd", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_int32(const int32_t number)
{
    /* define how many digits there are in number */
    const int digits = (number < 0) ? (int)(log10(number) + 2)
                                    : (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%d", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_int16(const int16_t number)
{
    /* define how many digits there are in number */
    const int digits = (number < 0) ? (int)(log10(number) + 2)
                                    : (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%hd", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_int8(const int8_t number)
{
    /* define how many digits there are in number */
    const int digits = (number < 0) ? (int)(log10(number) + 2)
                                    : (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%hd", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_int(const int number)
{
    /* define how many digits there are in number */
    const int digits = (number < 0) ? (int)(log10(number) + 2)
                                    : (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%d", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

e_string_t e_string_from_uint(const unsigned int number)
{
    /* define how many digits there are in number */
    const int digits = (int)(log10(number) + 1);
    char* cstr = calloc(digits, sizeof(uint8_t));
    snprintf(cstr, digits, "%u", number);

    e_string_t result = e_string_from_cstr(cstr);

    free(cstr);

    return result;
}

include/e_string.h
/* Copyright (c) 2023, diogoefl
 * SPDX-License-Identifier: BSD-3-Clause
 * See LICENSE file at this project root for more detailed information
 */

/* e_string header 
 *
 * this module implements UTF-8 growable strings array with size tracking.
 * api inspiration is based on Rust std::String and data structure is also 
 * inspired by Luca Sas string handling library exposed at ACCU 2021 conference
 * talk available at: https://www.youtube.com/watch?v=QpAhX-gsHMs&t=3009s
 * 
 * usage: add #include "e_string.h" to your file
 */

#ifndef E_STRING_H
#define E_STRING_H

#include <stddef.h>
#include <stdint.h>
#include <stdbool.h>

/* NAMESPACE E_STRING *********************************************************/

/* e_string struct
 *
 * this is the basic data structure to keep track on string size and capacity.
 * data will be stored always UTF-8 encoding, and functionality for conversion
 * can be used to retrieve cstr (NUL terminated strings).
 * 
 * PODs definition
 *   - data_length: defines the current length for the stored string
 *   - buffer_capacity: defines the amount of stored memory in the buffer
 *   - data: defines pointer to the start of the UTF-8 string data
 * 
 * length and capacity are stored as size_t to allow the maximum amount the
 * system can store - which is defined in SIZE_MAX .
 *
 * this is a growable string, therefore the buffer_capacity can be longer than
 * the data_length, but if for some reason the data_length is longer than the
 * buffer_capacity - an undefined behaviour can occur during the usage.
 *
 * the data is stored as UTF-8 strings, user must convert there string to UTF-8
 * compatible data prior to creating e_string_t.
 * functions for conversion are available at this library as well.
 */
typedef struct e_string
{
    size_t data_length;
    size_t buffer_capacity;
    uint8_t * data;
} e_string_t;

/* e_string errno macros
 *
 * this is a simple data that defines typical error that this library can exit.
 * all the errno used for e_string use the composition of the following numbers:
 *  - -5: defines e_lib prefix for errno to prevent conflict with c_errno
 *  -  1: defines e_string library
 *  -  0: defines specific errno
 * 
 * as an example -511 means an invalid buffer size (E_STRING_INVALID_BUFFER)
 * 
 * e_string_errno are negative in case of error to allow for easier comparison
 * on if statements.
 */
typedef int e_string_errno_t;
#define E_STRING_INVALID_BUFFER -511
#define E_STRING_INVALID_UTF8   -512
#define E_STRING_ERROR           false   /* 0 */
#define E_STRING_SUCCESS         true    /* 1 */


/* constructors
 * use this group of functions to create e_string_t data.
 */

/* e_string_from macro
 *
 * this macro delegates to the specialized constructor based on input type.
 */
#define e_string_from(X) _Generic((X),                        \
                         const char*:  e_string_from_cstr,    \
                         uint64_t:     e_string_from_uint64,  \
                         uint32_t:     e_string_from_uint32,  \
                         uint16_t:     e_string_from_uint16,  \
                         uint8_t:      e_string_from_uint8,   \
                         int64_t:      e_string_from_int64,   \
                         int32_t:      e_string_from_int32,   \
                         int16_t:      e_string_from_int16,   \
                         int8_t:       e_string_from_int8,    \
                         int:          e_string_from_int,     \
                         unsigned int: e_string_from_uint,    \
                         )(X)

/* e_string_from_cstr
 *
 * this function allows the user to create a e_string based on cstr.
 * it expects the user to provide a UTF-8 compliant cstr.
 */
e_string_t e_string_from_cstr(const char* cstr);

/* e_string_from_uint64
 *
 * this function allows the user to create based on unsigned 64bit integer.
 */
e_string_t e_string_from_uint64(const uint64_t number);

/* e_string_from_uint32
 *
 * this function allows the user to create based on unsigned 32bit integer.
 */
e_string_t e_string_from_uint32(const uint32_t number);

/* e_string_from_uint16
 *
 * this function allows the user to create based on unsigned 16bit integer.
 */
e_string_t e_string_from_uint16(const uint16_t number);

/* e_string_from_uint8
 *
 * this function allows the user to create based on unsigned 8bit integer.
 */
e_string_t e_string_from_uint8(const uint8_t number);

/* e_string_from_int64
 *
 * this function allows the user to create based on signed 64bit integer.
 */
e_string_t e_string_from_int64(const int64_t number);

/* e_string_from_int32
 *
 * this function allows the user to create based on signed 32bit integer.
 */
e_string_t e_string_from_int32(const int32_t number);

/* e_string_from_int16
 *
 * this function allows the user to create based on signed 16bit integer.
 */
e_string_t e_string_from_int16(const int16_t number);

/* e_string_from_int8
 *
 * this function allows the user to create based on signed 8bit integer.
 */
e_string_t e_string_from_int8(const int8_t number);

/* e_string_from_int
 *
 * this function allows the user to create based on signed integer.
 */
e_string_t e_string_from_int(const int number);

/* e_string_from_uint
 *
 * this function allows the user to create based on unsigned integer.
 */
e_string_t e_string_from_uint(const unsigned int number);

/* data access and validation
 * use this group of functions to validate e_string_t data and safe access.
 */

/* e_string_validate
 *
 * this function allows the user to validate if the string is valid, either by
 * checking if the data is UTF-8 compatible, or if the buffer and data is of
 * compatible size.
 *
 * it does not change the given data, nor it fix invalid structures.
 */
e_string_errno_t e_string_validate(const e_string_t* string);




/* NAMESPACE E_STRING_UTF8 ****************************************************/


/* e_string_utf8_is_ascii
 *
 * checks if data is compliant to US ASCII on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * US ASCII is represented in e_string_t as 1 byte (8 bit) with the value
 * ranging from 0x00 upto 0x7F. 
 *
 * if sanity check is true, the following items from 0x00 to 0x1F are considered
 * invalid, as they can represent conflicts for usage and should not be
 * inside of a e_string_t which attempts to represent simple text strings.
 * if you need a string with this type of characters, please use char* instead.
 *   - 0x00 NUL
 *   - 0x01 SOH
 *   - 0x02 STX
 *   - 0x03 ETX
 *   - 0x04 EOT
 *   - 0x05 ENT
 *   - 0x06 ACK
 *   - 0x07 \a
 *   - 0x08 \b
 *   - 0x0B VT
 *   - 0x0C \f
 *   - 0x0E SO
 *   - 0x0F SI
 *   - 0x10 DLE
 *   - 0x11 DC1
 *   - 0x12 DC2
 *   - 0x13 DC3
 *   - 0x14 DC4
 *   - 0x15 NAK
 *   - 0x16 SYN
 *   - 0x17 ETB
 *   - 0x18 CAN
 *   - 0x19 EM
 *   - 0x1A SUB
 *   - 0x1B \e
 *   - 0x1C FS
 *   - 0x1D GS
 *   - 0x1E RS
 *   - 0x1F US
 *   - 0x7F DEL
 *
 * ranging from 0x00 to 0x1F, the following items are valid for e_string_t,
 * regardless of sanity check value.
 *   - 0x09 \t
 *   - 0x0A LF
 *   - 0x0D CR
 */
bool e_string_utf8_is_ascii(const uint8_t u8_byte, const bool sanity);

/* e_string_utf8_is_non_overlong
 *
 * checks if data is compliant to non-overlong 2 bytes on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * non-overlong is represented in e_string_t as 2 bytes (16 bit) with the value
 * of the first byte ranging from 0xC2 upto 0xDF and the second byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is non-overlong you need both bytes for the correct output.
 */
bool e_string_utf8_is_non_overlong(const uint8_t u8_1st_byte,
                                   const uint8_t u8_2nd_byte);


/* e_string_utf8_is_excluding_overlong
 *
 * checks if data is compliant to excluding overlong 3 bytes on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * excluding overlong is represented in e_string_t as 3 bytes (24 bit) with the
 * value of the first byte being 0xE0, the second byte ranging from 0xA0 upto
 * 0xBF and the third byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is excluding overlong you need all bytes for the correct output.
 */
bool e_string_utf8_is_excluding_overlong(const uint8_t u8_1st_byte,
                                         const uint8_t u8_2nd_byte,
                                         const uint8_t u8_3rd_byte);


/* e_string_utf8_is_straight_3byte
 *
 * checks if data is compliant to straight 3-byte on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * straight 3-byte is represented in e_string_t as 3 bytes (24 bit) with the
 * value of the first byte ranging from 0xE1 upto 0xEC or from 0xEE upto 0xEF,
 * the second byte ranging from 0x80 upto 0xBF and the third byte ranging from
 * 0x80 upto 0xBF.
 *
 * if the codepoint is straight 3-byte you need all bytes for the correct output.
 */
bool e_string_utf8_is_straight_3byte(const uint8_t u8_1st_byte,
                                     const uint8_t u8_2nd_byte,
                                     const uint8_t u8_3rd_byte);


/* e_string_utf8_is_excluding_surrogates
 *
 * checks if data is compliant to excluding surrogates on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * excluding surrogates is represented in e_string_t as 3 bytes (24 bit) with 
 * the value of the first byte is 0xED, the second byte ranging from 0x80 upto
 * 0x9F and the third byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is excluding surrogates you need all bytes for the correct output.
 */
bool e_string_utf8_is_excluding_surrogates(const uint8_t u8_1st_byte,
                                           const uint8_t u8_2nd_byte,
                                           const uint8_t u8_3rd_byte);


/* e_string_utf8_is_plane_1to3
 *
 * checks if data is compliant to planes 1-3 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * planes 1-3 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte is 0xF0, the second byte ranging from 0x90 upto
 * 0xBF, the third byte ranging from 0x80 upto 0xBF and the fourth byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is planes 1-3 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_1to3(const uint8_t u8_1st_byte,
                                 const uint8_t u8_2nd_byte,
                                 const uint8_t u8_3rd_byte,
                                 const uint8_t u8_4th_byte);


/* e_string_utf8_is_plane_4to15
 *
 * checks if data is compliant to planes 4-15 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * planes 4-15 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte ranging from 0xF1 upto 0xF3, the second byte
 * ranging from 0x80 upto 0xBF, the third byte ranging from 0x80 upto 0xBF and
 * the fourth byte ranging from 0x80 upto 0xBF.
 *
 * if the codepoint is planes 4-15 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_4to15(const uint8_t u8_1st_byte,
                                  const uint8_t u8_2nd_byte,
                                  const uint8_t u8_3rd_byte,
                                  const uint8_t u8_4th_byte);


/* e_string_utf8_is_plane_16
 *
 * checks if data is compliant to plane 16 on UTF-8 encoding, 
 * based on definitions of 
 * https://www.w3.org/International/questions/qa-forms-utf-8
 * 
 * plane 16 is represented in e_string_t as 4 bytes (32 bit) with 
 * the value of the first byte is 0xF4, the second byte ranging from 0x80 upto
 * 0x8F, the third byte ranging from 0x80 upto 0xBF and the fourth byte ranging
 * from 0x80 upto 0xBF.
 *
 * if the codepoint is plane 16 you need all bytes for the correct output.
 */
bool e_string_utf8_is_plane_16(const uint8_t u8_1st_byte,
                               const uint8_t u8_2nd_byte,
                               const uint8_t u8_3rd_byte,
                               const uint8_t u8_4th_byte);


#endif /* E_STRING_H */